server/src/main/java/org/elasticsearch/common/logging/ChunkedLoggingStream.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * An {@link OutputStream} which Gzip-compresses the written data, Base64-encodes it, and writes it in fixed-size chunks to a logger. This
+ * is useful for debugging information that may be too large for a single log message and/or which may include data which cannot be
+ * recorded faithfully in plain-text (e.g. binary data or data with significant whitespace).
+ */
+public class ChunkedLoggingStream extends OutputStream {
+
+    static final int CHUNK_SIZE = ByteSizeUnit.KB.toIntBytes(2);
+
+    /**
+     * Create an {@link OutputStream} which Gzip-compresses the written data, Base64-encodes it, and writes it in fixed-size (2kiB) chunks
+     * to the given logger. If the data fits into a single chunk then the output looks like this:
+     *
+     * <pre>
+     * $PREFIX (gzip compressed and base64-encoded): H4sIAAAAA...
```

Comment:
Does this mean the new lag output will not be human-readable text anymore? I am not keen about the fact that semi-manual reconstruction will be required.

Reply:
Yes, that's deliberate. We have scripts that process hot threads output, but they don't work so well if it gets mangled (e.g. whitespace lost) on the way to us. I also intend to try using this for #38939, which would involve capturing non-ASCII data that won't translate well in logs. Again, this is only really for debugging, and it's pretty trivial to decode this (e.g. `:%!gzip -d|base64 -D` in vi).

Reply:
I see. I have seen the LagDetector's output in issues and it is quite useful in a human-readable form. If we change it, I would appreciate if you can:

* Modify the [relevant documentation](https://www.elastic.co/guide/en/elasticsearch/reference/8.6/cluster-fault-detection.html#_diagnosing_lagging_nodes) to mention an easy way to decode it into human-readable format (maybe with the `vi` command you mentioned, or better yet some terminal commands to do it).
* Maybe also write the same command in the final line logged, so it is easy for a human to try to decode it to a human-readable form.

I understand that for debugging purposes, there may be tools that will do all this automatically, but I feel that it will be useful also for less experienced humans to get the human-readable output.

Reply:
Just because I feel a bit uneasy on this, @tlrx feel free to comment as well. I'd feel better if at least 2 people are fine with the encoded output :)

Reply:
Would an alternative be to store the lag log in a separate log file? E.g., configure a log4j configuration so the LagDetector's log goes into a separate file?

Reply:
Fair point, see a93504724c1c2a7325c2e624302b3406e4bddb7d.

If we put it in a separate file it would be much less likely to be shared with us. For instance, in Cloud we would have to change the whole log indexing pipeline to handle the new file, and experience shows that on-prem folks are sometimes pretty bad at sharing the info we need even when it's in a standard place and format.

Reply:
Thanks for the explanation, that makes sense. I like the documentation changes.